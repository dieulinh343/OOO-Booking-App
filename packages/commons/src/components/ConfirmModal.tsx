import React, { useState } from 'react';
import Modal, { FooterType } from './Modal';

export function ConfirmModal({
  waitConfirm,
  onConfirm,
  onCancel,
  onModalClose,
  headerText,
  bodyText,
  primaryButtonText,
  isDangerous = false,
  centered = false,
  ...rest
}: {
  waitConfirm?: boolean;
  onConfirm: () => void;
  onCancel: () => void;
  onModalClose: () => void;
  headerText: string;
  bodyText: string;
  primaryButtonText: string;
  isDangerous: boolean;
  centered: boolean;
  rest?: object;
}) {
  const [submitting, setSubmitting] = useState(false);

  return (
    <Modal
      id="modal-confirm"
      centered={centered}
      closable={false}
      size="small"
      headerText={headerText || 'Confirmation'}
      body={(
        <p>{bodyText}</p>
      )}
      primaryButtonText={submitting ? 'SUBMITTING...' : (primaryButtonText || 'YES')}
      onClickPrimaryButton={async () => {
        if (waitConfirm) {
          setSubmitting(true);
          await onConfirm();
        } else {
          onConfirm();
        }

        onModalClose();
      }}
      secondaryButtonText="NO"
      onClickSecondaryButton={() => {
        onCancel && onCancel();
        onModalClose();
      }}
      primaryButtonVariant={isDangerous ? 'negative' : 'primary'}
      footerType={FooterType.DOUBLE}
      disablePrimaryButton={submitting}
      disableSecondaryButton={submitting}
      {...rest}
    />
  );
}
export default ConfirmModal;
