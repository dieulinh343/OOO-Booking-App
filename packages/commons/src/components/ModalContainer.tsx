import React from 'react';

interface ModalContainerProps<T> {
  modalsMap: Record<string, React.FC<any>>;
  modalProps: T;
  handleShowModal?: (value: string | null) => void;
  handleCloseModal?: () => void;
}

export function ModalContainer<T extends {
  displayModal: string | null;
  onModalClose?: ((e: EventTarget) => void) | null;
}>({
  modalsMap,
  modalProps,
  handleCloseModal,
  // Keep this for backward compatibility
  handleShowModal,
}: ModalContainerProps<T>) {
  // Hide displaying modal
  const handleModalClose = () => {
    if (handleCloseModal) {
      handleCloseModal();
      return;
    }

    // Backward compatibility
    if (handleShowModal) {
      handleShowModal(null);
    }
  };

  const { displayModal, onModalClose, ...rest } = modalProps;
  const RenderedModal: React.FC<any> | null = (
    displayModal === null
      ? null
      : modalsMap[displayModal]
  );
  const currentProps = {
    ...rest,
    onModalClose: (e: EventTarget) => {
      handleModalClose();
      onModalClose && onModalClose(e);
    },
  };

  return RenderedModal
    ? <RenderedModal {...currentProps} />
    : null;
}

export default ModalContainer;
