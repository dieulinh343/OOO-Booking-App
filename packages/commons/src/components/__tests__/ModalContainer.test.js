import React from 'react';
import { shallow } from 'enzyme';
import { ModalContainer } from '../ModalContainer';
import ConfirmModal from '../ConfirmModal';

describe('components/ModalContainer', () => {
  let wrapper;
  let props;

  const setup = () => {
    wrapper = shallow(
      <ModalContainer
        {...props}
      />);
  };

  beforeEach(() => {
    props = {
      modalsMap: {
        confirm: ConfirmModal,
      },
      modalProps: {
        displayModal: 'confirm',
        onModalClose: jest.fn(),
        onConfirm: jest.fn(),
        bodyText: 'sample body text',
      },
      handleShowModal: jest.fn(),
      handleCloseModal: jest.fn(),
    };
  });

  it('should render modal container successfully', () => {
    setup();
    expect(wrapper.length).toEqual(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('should trigger handleCloseModal (if exists) when onModalClose is called', () => {
    setup();
    wrapper.props().onModalClose();
    expect(props.handleCloseModal).toHaveBeenCalled();
    expect(props.handleShowModal).not.toHaveBeenCalled();
  });

  it('should trigger handleShowModal (if exists) when onModalClose is called', () => {
    delete props.handleCloseModal;
    setup();
    wrapper.props().onModalClose();
    expect(props.handleShowModal).toHaveBeenCalledWith(null);
  });

  it('should show nothing when displayModal props is empty', () => {
    props = {
      ...props,
      modalProps: {
        ...props.modalProps,
        displayModal: null,
      },
    };
    setup();
    expect(wrapper).toMatchSnapshot();
  });
});
