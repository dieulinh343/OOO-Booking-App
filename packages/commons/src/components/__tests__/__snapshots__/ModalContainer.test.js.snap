// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ModalContainer should render modal container successfully 1`] = `
<ConfirmModal
  bodyText="sample body text"
  onConfirm={[MockFunction]}
  onModalClose={[Function]}
/>
`;

exports[`components/ModalContainer should show nothing when displayModal props is empty 1`] = `""`;
